### ISSUE TITLE
ggplot2

### ISSUE THREAD

**Issue by Godisemo** [tbreloff/Plots.jl#201]
Would it be hypothetically possible to add the R package ggplot2 as a backend? Compare with the julia package for pyplot. Which steps would be necessary to make this reality?

**Comment by lobingera** [comment]
Have you looked at Gadfly? It's supposed to be 'close' to ggplot2. To make this (ggplot2 as backend) happen, an efficient method to call R from julia would be needed.

**Comment by lobingera** [comment]
... and seems to be existing with https://github.com/JuliaStats/RCall.jl

**Comment by Godisemo** [comment]
I currently use gadfly, but for publication quality plots I still prefer
ggplot2. There seems to be a Julia to R package called Rif.jl, but I have
never tried it.

På 25 april 2016 till 12:35:05, Andreas Lobinger (notifications@github.com)
skrev:

**Comment by lobingera** [comment]
Rcall seems to be the continuation of Rif.jl
Just for curiosity, what is the publication quality difference between ggplot2 and Gadfly on the same plot? Could you do an example?

**Comment by tbreloff** [comment]
Yes the answer is that it could be made into a backend through RCall, but I'm not convinced that it's worth the effort. 

I have a feeling that a big draw of ggplot2 is the style (or theme)... Background, grid lines color, etc. 

I want thinking a couple weeks ago that it would be really nice to have a 'theme' keyword argument that would change many default values all at once to look like native matplotlib, ggplot2, etc. if this was available, I wonder if you would still care about calling into R. 

Please post examples of what this backend enables so there's justification to the request.

**Comment by Evizero** [comment]
I do admin that ggplot2 is still and clearly the most visually pleasing plotting system out there. a couple of gr plots come close though.

**Comment by joshday** [comment]
ggplot2 has some great features for plotting spatial data (http://docs.ggplot2.org/current/geom_map.html), but maybe that should be its own issue.

**Comment by Godisemo** [comment]
I could try to make some examples of why I think ggplot makes better publication quality plots than for example gadfly, though I believe ggplot is considered to be the best plotting library by a wide audience for a lot of reasons.

As @tbreloff, my concern is also whether it would be worth the effort. Is the current workflow good enough? That is: getting simulation results or similar in julia, plotting with for example gadfly, and when time comes for publication export data and import to R and plot with ggplot.

**Comment by tbreloff** [comment]
> I could try to make some examples of why I think ggplot makes better publication quality plots than for example gadfly

Please do.  The thing is, if we can exactly recreate a ggplot2 plot without calling into R... why bother.  If you give me a few examples to work with, I might be able to show you how to reproduce them in Plots (or at least it will give me direction to improve the package)

>  and when time comes for publication export data and import to R and plot with ggplot.

If this is required, I'd be a sad panda...

**Comment by tbreloff** [comment]
A preview of something I'm working on:

![tmp](https://cloud.githubusercontent.com/assets/933338/14788114/835312c8-0ad4-11e6-8ab1-fddbcc0a2354.png)

by allowing more fine-tuned control of individual plot colors, I could easily add a method like:

```julia
set_theme(:ggplot2)
```
which would update the default values for all these colors to produce the sort of look that you see above.

Currently I have "sub-background" values (for legend, inside the grid, and outside the grid) that default to `:match`, which means they take the main background color unless overridden.  Same for foreground colors (legend, axis marks, tick text, and borders).

Thoughts/ideas?

It should be very straightforward to add themes, and likely can be added on the fly by the user (as it would just be updating the dictionary holding default arguments).

**Comment by diegozea** [comment]
Things needed for publications outside the theme issue:

- [ ] Select font (i.e. PLoS ask for Arial, Times or Symbol... Setting Arial in ggplot2 isn't easy)
- [ ] Determine font size (i.e. PLoS ask for 8 to 12 points)
- [ ] Determine the point size of lines (i.e. PLoS ask for a minimum of 0.5 points)
- [ ] Export TIFF and EPS (several journals ask for tiff or eps files)
- [ ] Determine quality/resolution (dpi) (in general journals ask for 300 dpi or 600 dpi)  

Figure requirements/guidelines:
* [PLoS](http://journals.plos.org/plosone/s/figures#loc-requirements-at-a-glance)
* [PubMed Central](http://www.ncbi.nlm.nih.gov/pmc/pub/filespec-images/)

**Comment by tbreloff** [comment]
`add_theme`/`set_theme` are implemented.  It was pretty simple.  It'll only work right with the PyPlot backend, and only once I merge in the `tb_pyplot` branch.  Here's a sample:

![tmp](https://cloud.githubusercontent.com/assets/933338/14789270/bfb28280-0ad9-11e6-99ac-1092c1a53cdb.png)

**Comment by tbreloff** [comment]
See: 77c2d7d846f27246b3420f7eca942fbea5001560

**Comment by tbreloff** [comment]
>  Select font (i.e. PLoS ask for Arial, Times or Symbol... Setting Arial in ggplot2 isn't easy)
 Determine font size (i.e. PLoS ask for 8 to 12 points)

FYI... you can set font attributes using the `Plots.Font` type:

![tmp](https://cloud.githubusercontent.com/assets/933338/14789576/3340357a-0adb-11e6-947a-3ee4f3220a4d.png)

**Comment by diegozea** [comment]
That was more easy than with ggplot2 ;)

**Comment by Evizero** [comment]
I like where this issue is headed

**Comment by swt30** [comment]
How does this tie in with matplotlib's built-in styles or other packages that change the matplotlib rc parameters, such as seaborn? I've been trying to change the look of the outputs from Plots by doing something like
```
using Plots, PyCall
@pyimport matplotlib.style as style
style.use("bmh")
plot(...)
```
or
```
using Plots, PyCall
@pyimport seaborn as sns
plot(...)
sns.despine()
```
but not really having any luck.

**Comment by tbreloff** [comment]
It doesn't tie in with them.  Plots doesn't really use the defaults of matplotlib, so something like seaborn that depends on setting the defaults won't work as expected.  It's technically feasible that you could ask Plots to grab default values from matplotlib, but that doesn't seem like it's worth the trouble... I'd rather add a theme that all backends can use.

**Comment by swt30** [comment]
Makes sense for a meta-plotting package, I guess. It's a just a shame because I love using the API of Plots such much, but then wanted to go in and make tweaks to the plot before publication. Is there a way to instead tell Plots not to apply any styling at all so I can deal with that bit?

**Comment by tbreloff** [comment]
> make tweaks to the plot before publication

You can still call PyPlot functions if you want.

>  Is there a way to instead tell Plots not to apply any styling at all so I can deal with that bit?

This is a pretty slipperly slope, though I see the draw if you want to use something like seaborn for themes.  I would prefer that anything you want is available from Plots, and otherwise you can update the plot after-the-fact directly through the backend.

If you're having trouble getting it looking the way you want, please post specific examples and I'll try to help.  Bypassing only some of the Plots processing and method calls will likely be a tricky thing to get right...

**Comment by swt30** [comment]
After a little tinkering, I'm really impressed with how simple it is to just go and do things like turn off the grid directly through Plots without having to even touch PyPlot at all. Making themes that all backends can use is definitely the way to go :+1: I hit one little snag but I'll post it as a separate issue.

**Comment by pkofod** [comment]
Should themes have a seperate "variety" keyword, or should they just be `solarized1`, `solarized2` etc?

**Comment by tbreloff** [comment]
My vote would probably be to have a main 'solarized' theme, and then
'solarized_xxx' which uses 'base = solarized'. (Assuming the themes share
some attributes)

**Comment by pkofod** [comment]
I think I agree. For solarized for example, it makes sense to have a dark and light theme, but for others it might not.

**Comment by Evizero** [comment]
Not sure if this has been mentioned before. but [ggthemes](https://github.com/jrnold/ggthemes#examples) is a pretty nice place to get inspiration

**Comment by Evizero** [comment]
emphasis on inspiration. It's GPL-2 after all

**Comment by pkofod** [comment]
Someone mentioned it (I thougt it was you?) Luckily the color definitions are often available at the sources, so I don't have to base it directly off of ggthemes, I just plan to use it to find the sources :) But thanks.

**Comment by diegozea** [comment]
NOTE: *tiff* and other formats can be saved from the PyPlot's *gui* :)

**Comment by crbinz** [comment]
Lots of good stuff in this issue, glad I found it. Is anyone actively working on themes, either original or reproductions of, e.g., the seaborn matplotlib themes? Is the thought that all themes will live in this package, or some separate one?

**Comment by tbreloff** [comment]
PlotThemes.jl

**Comment by pkofod** [comment]
Yes, that's PlotThemes.jl territory. Unfortunately I've had to snap back to reality and work on my thesis the past few months, so it's been standing entirely still. However, PRs are welcome. Though, I should resume work (on *.jl) relatively soon.

**Comment by tbreloff** [comment]
I need to add a "themes" page to the docs so that people know about this feature (and PlotThemes.jl)
